nebula_ros/src/hesai/decoder_wrapper.cpp

```diff
@@ -0,0 +1,317 @@
+  auto new_driver = std::make_shared<drivers::HesaiDriver>(new_config, calibration_cfg_ptr_);
+  driver_ptr_ = new_driver;
+  sensor_cfg_ = new_config;
+}
+
+void HesaiDecoderWrapper::OnCalibrationChange(
+    const std::shared_ptr<const nebula::drivers::HesaiCalibrationConfigurationBase>& new_calibration)
+{
+  std::lock_guard lock(mtx_driver_ptr_);
+  auto new_driver = std::make_shared<drivers::HesaiDriver>(sensor_cfg_, new_calibration);
+  driver_ptr_ = new_driver;
+  calibration_cfg_ptr_ = new_calibration;
+  calibration_file_path_ = calibration_cfg_ptr_->calibration_file;
+}
+
+rcl_interfaces::msg::SetParametersResult HesaiDecoderWrapper::OnParameterChange(const std::vector<rclcpp::Parameter>& p)
+{
+  using namespace rcl_interfaces::msg;
+
+  std::string calibration_path = "";
+
+  // Only one of the two parameters is defined, so not checking for sensor model explicitly here is fine
+  bool got_any = get_param(p, "calibration_file", calibration_path) | get_param(p, "correction_file", calibration_path);
+  if (!got_any)
+  {
+    return rcl_interfaces::build<SetParametersResult>().successful(true).reason("");
+  }
+
+  if (!std::filesystem::exists(calibration_path))
+  {
+    auto result = SetParametersResult();
+    result.successful = false;
+    result.reason = "The given calibration path does not exist, ignoring: '" + calibration_path + "'";
+    return result;
+  }
+
+  auto get_calibration_result = GetCalibrationData(calibration_path, true);
+  if (!get_calibration_result.has_value())
+  {
+    auto result = SetParametersResult();
+    result.successful = false;
+    result.reason = (std::stringstream() << "Could not change calibration file to '" << calibration_path
+                                         << "': " << get_calibration_result.error())
+                        .str();
+    return result;
+  }
+
+  OnCalibrationChange(get_calibration_result.value());
+  RCLCPP_INFO_STREAM(logger_, "Changed calibration to '" << calibration_cfg_ptr_->calibration_file << "'");
+  return rcl_interfaces::build<SetParametersResult>().successful(true).reason("");
+}
+
+HesaiDecoderWrapper::get_calibration_result_t
+HesaiDecoderWrapper::GetCalibrationData(const std::string& calibration_file_path, bool ignore_others)
+{
+  std::shared_ptr<drivers::HesaiCalibrationConfigurationBase> calib;
+
+  if (sensor_cfg_->sensor_model == drivers::SensorModel::HESAI_PANDARAT128)
+  {
+    calib = std::make_shared<drivers::HesaiCorrection>();
+  }
+  else
+  {
+    calib = std::make_shared<drivers::HesaiCalibrationConfiguration>();
+  }
+
+  bool hw_connected = hw_interface_ != nullptr;
+  std::string calibration_file_path_from_sensor;
+
+  {
+    int ext_pos = calibration_file_path.find_last_of('.');
+    calibration_file_path_from_sensor = calibration_file_path.substr(0, ext_pos);
+    // TODO: if multiple different sensors of the same type are used, this will mix up their calibration data
+    calibration_file_path_from_sensor += "_from_sensor";
```

Comment:
I agree that the SN would be the best approach. I'll see if it is viable to include it in this PR!

Reply:
Because this code has to be able to function even without a sensor connected, curently IP is our easiest option.
Otherwise we would need S/N or MAC address as a parameter and that's out of scope for this PR.

Reply:
Addressed in [b2552dc](https://github.com/tier4/nebula/pull/127/commits/b2552dc14420cf1969a7ae5ba2f0ca720c3b8112), appended IP address to filename (e.g. `OT128_from_sensor.csv` :arrow_right: `OT128_from_sensor_192.168.1.201.csv`)